src/lib.rs

```diff
@@ -277,88 +281,57 @@ enum IpfsEvent {
     Exit,
 }
 
-/// Configured Ipfs instace or value which can be only initialized.
-pub struct UninitializedIpfs<Types: IpfsTypes> {
-    repo: Repo<Types>,
-    span: Span,
-    keys: Keypair,
-    options: IpfsOptions,
-    repo_events: Receiver<RepoEvent>,
-}
-
-impl<Types: IpfsTypes> UninitializedIpfs<Types> {
+impl<Types: IpfsTypes> Ipfs<Types> {
     /// Configures a new UninitializedIpfs with from the given options and optionally a span.
     /// If the span is not given, it is defaulted to `tracing::trace_span!("ipfs")`.
     ///
     /// The span is attached to all operations called on the later created `Ipfs` along with all
     /// operations done in the background task as well as tasks spawned by the underlying
     /// `libp2p::Swarm`.
-    pub async fn new(options: IpfsOptions, span: Option<Span>) -> Self {
+    pub fn new(options: IpfsOptions, span: Option<Span>) -> Self {
         let repo_options = RepoOptions::from(&options);
         let (repo, repo_events) = create_repo(repo_options);
-        let keys = options.keypair.clone();
         let span = span.unwrap_or_else(|| tracing::trace_span!("ipfs"));
+        let to_task = channel::<IpfsEvent>(1).0; // FIXME: dummy
 
-        UninitializedIpfs {
-            repo,
+        Ipfs {
+            repo: Arc::new(repo),
             span,
-            keys,
             options,
-            repo_events,
+            repo_events: Some(repo_events),
+            to_task,
         }
     }
 
-    pub async fn default() -> Self {
-        Self::new(IpfsOptions::default(), None).await
+    pub fn default() -> Self {
+        Self::new(IpfsOptions::default(), None)
     }
 
     /// Initialize the ipfs node. The returned `Ipfs` value is cloneable, send and sync, and the
     /// future should be spawned on a executor as soon as possible.
-    pub async fn start(self) -> Result<(Ipfs<Types>, impl Future<Output = ()>), Error> {
+    pub async fn start(&mut self) -> Result<impl Future<Output = ()>, Error> {
```

Comment:
The story behind of `UninitializedIpfs` was probably to originally work around some Send or Sync bounds related to some type which is no longer an issue however I think the real benefit of `UninitalizedIpfs` was that there was no `start` on `Ipfs` which would be called multiple times as the move semantics of the lang give rise to nicer APIs.

What does one do with an `Ipfs` value if the `start` fails?

Could the `start` be moved to IpfsOptions...? As a `build` method or something? Or maybe in place of `new`?

Reply:
> What does one do with an Ipfs value if the start fails?

same applies to `UninitializedIpfs`, no? I'm all for improving the naming here, I'll think about it

Reply:
Well `UninitializedIpfs` is a worthless value in itself except for starting, and `start` consumed itself, so you don't have a value left.

Reply:
while working on ipfs-substrate I had a case where it was easy to spawn tasks, but you had no control over what they were returning, making the setup practical for long-running tasks; in this case the benefit is that `Ipfs` can be created (and kept) in a non-`async` context, while the background task can be concealed away; actually, since `Ipfs` is `Clone`, this can still consume `self`, would that be better?